### DISCUSSION TITLE
Unable to access file from supabase storage

### DISCUSSION THREAD

**Discussion by rmrbytes** [supabase/supabase#9829] [Questions] [answered]
Hello:

I am creating an editing type of app and store the contents typed by the user as a text file in Storage. I am able to create the file (I can see it in the storage console). However, when I try to access it, the data is null.

My code is:

```
const filename = 'physics/content.txt'  //  I set this to the correct file
const { data, error } = await supabase
    .storage
    .from('tutorials')  // the bucket name
    .download(_filename)

// check for error    
if (error) {
    console.log('storage error', error) // no error appears
    throw error
}
console.log(data) // returns null
```

Strangely, even if I give a wrong filename, no error is captured though the browser console shows a 400 error.

I tried making the bucket public but that too does not get the contents of the file.

Thanks

**Comment by GaryAustin1** [comment]
You will need to set select policy on the storage.objects table.  Public bucket only means you can use the public url for the file to download it, it does not impact API operations.  

https://supabase.com/docs/guides/storage#add-security-rules

**Reply by rmrbytes** [reply]
Thanks @GaryAustin1 . I forgot to mention (my bad) that I do have the policy set. That is how the creation bit is working. The policy I have for all: select, insert, update and delete is:

`((bucket_id = 'tutorials'::text) AND (role() = 'authenticated'::text))`

And only authenticated users are able to access the page

**Reply by GaryAustin1** [reply]
The policy must include USING (for Select for instance) and WITH CHECK for upload.

You say authenticated users can access the page, but also say you get a 400 error?

Another thing would be to make sure the filename does not include the bucket name.  

You may need to show the URL generated by your request to get more help.

**Reply by rmrbytes** [reply]
Thanks again @GaryAustin1 for the help. This is the URL

`https://dystldfsammjyjzozzlq.supabase.co/storage/v1/object/sign/tutes/52ce9640-709e-45d7-9eb1-4e5783308ae1/52ce9640-709e-45d7-9eb1-4e5783308ae1.html?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1cmwiOiJ0dXRlcy81MmNlOTY0MC03MDllLTQ1ZDctOWViMS00ZTU3ODMzMDhhZTEvNTJjZTk2NDAtNzA5ZS00NWQ3LTllYjEtNGU1NzgzMzA4YWUxLmh0bWwiLCJpYXQiOjE2NjY3NTM5OTksImV4cCI6MTk4MjExMzk5OX0.2M-YIH1sjQYvp_ihqccOLJGbd503EhVkN1qqlvo0xGo`

In the app, I am creating a folder (uuid of my database record) and within that there will be multiple files. Each file will be a UUID of a corresponding record (for each student) having content (in html format)

**Reply by rmrbytes** [reply]
![Screenshot from 2022-10-26 08-53-53](https://user-images.githubusercontent.com/14859054/197927250-17b1ca54-cf77-4e75-b33a-a33783239574.png)

This is how the policy looks in the console. I did not understand what you meant by CHECK for upload. As I mentioned I am able to save files in storage using the same policy for INSERT. It is the SELECT policy that somehow does not seem to retrieve the file. Btw, the 400 error now does not appear.

**Reply by rmrbytes** [reply]
I even changed the policy to just `(bucket_id = 'tutes'::text)` and yet it shows null. There is some basic mistake that I am doing that is not showing up as an error when using the download method.

**Reply by GaryAustin1** [reply]
Various policies have either a USING policy (like select and update) or WITH CHECK policy (like insert).  Update can actually have both, but usually USING is good and WITH CHECK defaults to the USING policy for that.    USING is current record, WITH CHECK is new record.
Can you show your select policy?

**Reply by rmrbytes** [reply]
`((bucket_id = 'tutes'::text) AND (role() = 'authenticated'::text))`

**Reply by rmrbytes** [reply]
I tried removing the role condition and it still is null

**Reply by GaryAustin1** [reply]
That does not clarify if it is USING or WITH CHECK Policy...
https://supabase.com/docs/guides/storage

**Reply by rmrbytes** [reply]
I was checking the logs and this is what it shows

```
GET | 304 | 10.102.25.3 | 760069152742849e-BOM | /object/tutes/52ce9640-709e-45d7-9eb1-4e5783308ae1/52ce9640-709e-45d7-9eb1-4e5783308ae1.html | Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Safari/537.36
ISO Timestamp
2022-10-26T04:22:22.375Z

Metadata
[
  {
    "err": [],
    "req": [
      {
        "headers": [
          {
            "accept": "*/*",
            "cf_connecting_ip": "110.226.177.76",
            "cf_ipcountry": null,
            "cf_ray": "760069152742849e-BOM",
            "content_length": null,
            "content_type": null,
            "date": null,
            "host": "storage-api-lb-ap-south-1-ext.storage.supabase.com",
            "referer": "http://localhost:3000/",
            "user_agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Safari/537.36",
            "x_client_info": "supabase-js/2.0.4",
            "x_forwarded_proto": "https",
            "x_real_ip": null
          }
        ],
        "hostname": "storage-api-lb-ap-south-1-ext.storage.supabase.com",
        "method": "GET",
        "remoteAddress": "10.102.25.3",
        "remotePort": 7238,
        "url": "/object/tutes/52ce9640-709e-45d7-9eb1-4e5783308ae1/52ce9640-709e-45d7-9eb1-4e5783308ae1.html"
      }
    ],
    "prefix": null,
    "project": "dystldfsammjyjzozzlq",
    "expiresIn": null,
    "_7": null,
    "_5": null,
    "id": null,
    "level": "info",
    "res": [
      {
        "statusCode": 304
      }
    ],
    "name": null,
    "_4": null,
    "_0": null,
    "_2": null,
    "reqId": "760069152742849e-BOM",
    "sort_by": [],
    "responseTime": 87.03323101997375,
    "busboyOptions": [],
    "sortBy": [],
    "error": [],
    "_8": null,
    "tenantId": "dystldfsammjyjzozzlq",
    "search": null,
    "_1": null,
    "offset": null,
    "_6": null,
    "ascending": null,
    "results": [],
    "offser": null,
    "context": [
      {
        "host": "ip-10-102-62-210.ap-south-1.compute.internal",
        "pid": 27
      }
    ],
    "_3": null,
    "origObject": [],
    "next": null,
    "bucketId": null,
    "limit": null
  }
]
```

**Reply by rmrbytes** [reply]
> That does not clarify if it is USING or WITH CHECK Policy... https://supabase.com/docs/guides/storage

oh! ... but I just clicked on edit on the list of policies screen and this is what it shows. Let me investigate. Thanks

**Reply by GaryAustin1** [reply]
Your policy looks correct if you really have an authenticated user when you make the call.  You could retest with just the bucket name part to at least rule that out again.
You first code used 'tutorials" for the bucket, I assume it matches the bucket you are using?
I assume you can download the file from the UI and it has data.
Double check the filename in the url does not have an extra "folder" you don't expect in the url.

I'm out of ideas otherwise

**Reply by rmrbytes** [reply]
> You policy looks correct if you really have an authenticated user when you make the call. You could retest with just the bucket name part to at least rule that out again. You first code used 'tutorials" for the bucket, I assume it matches the bucket you are using? I assume you can download the file from the UI and it has data. Double check the filename in the url does not have an extra "folder" you don't expect in the url.
> 
> I'm out of ideas otherwise

Thanks for all the help @GaryAustin1. I am foxed with what finally worked. Will post it in the storage issues so that the team can comment.

**Comment by rmrbytes** [comment]
I deleted the policy from the console and ran the following using SQL Editor. Still it shows null

```
create policy "Logged Users to Store Tutes"
on storage.objects for select
using (
  bucket_id = 'tutes'
  and auth.role() = 'authenticated'
);
```

**Comment by rmrbytes** [comment, marked as answer]
Finally this worked. It is not related to policies.

```
const _filename = ....
const data  = await supabase
                .storage
                .from('tutes')
                .download(_filename)
```

Earlier I had put

```
const { data, error } = await supabase
              .storage
              .from('tutes')
              .download(_filename)
```
I am surprised.  

The version of supabase I am using is 2.0.4
